src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py

```diff
@@ -551,32 +548,30 @@ def _predict_with_probabilities(self, input_dataset, batch_size=None,
                 batch_data = batch.data[0]
                 size = batch_size
 
-            if batch_data.shape[0] < len(ctx):
-                ctx0 = ctx[:batch_data.shape[0]]
-            else:
-                ctx0 = ctx
-
-            z = self._model(batch_data).asnumpy()
-            predicted = z.argmax(axis=1)
+            ctx0 = ctx[:batch_data.shape[0]] if batch_data.shape[0] < len(ctx) else ctx
+            split_data = _mx.gluon.utils.split_and_load(batch_data, ctx_list=ctx0, even_split=False)
             classes = self.classes
-            
-            predicted_sa = _tc.SArray(predicted).apply(lambda x: classes[x])
-            
-            all_predicted[index : index + len(predicted_sa)] = predicted_sa
-            all_probabilities[index : index + z.shape[0]] = z
-            index += z.shape[0]
-            if index == dataset_size - 1:
-                done = True
-
-            cur_time = _time.time()
-            # Do not print process if only a few samples are predicted
-            if verbose and (dataset_size >= 5 
-                and cur_time > last_time + 10 or done):
-                print('Predicting {cur_n:{width}d}/{max_n:{width}d}'.format(
-                    cur_n = index,
-                    max_n = dataset_size,
-                    width = len(str(dataset_size))))
-                last_time = cur_time
+
+            for data in split_data:
+                z = self._model(data).asnumpy()
+                predicted = z.argmax(axis=1)
+                predicted_sa = _tc.SArray(predicted).apply(lambda x: classes[x])
+
+                all_predicted[index : index + len(predicted_sa)] = predicted_sa
```

Comment:
Shouldn't `len(predicted_sa)` and `z.shape[0]` be the same thing?

Using temporary a variable (e.x. `split_length`) could really help readability.

I don't understand why you're starting off by creating the full sized arrays and then doing this fancy indexing stuff. I think you could just start off (outside of the loop) with `all_predicted, all_probabilities = [], []` and then do:
```
all_predicted += predicted_sa
all_probabilities += z
```

Better yet since you're returning these things in SArray form, just use an [SArrayBuilder](https://apple.github.io/turicreate/docs/api/generated/turicreate.SArrayBuilder.html).

Reply:
Done.